### PR TITLE
Data-driven e2e tests

### PR DESCRIPTION
**Pull request by echeran**
This isn't complete (hence Draft PR), but I could use feedback, even on Rust basics (ex: code organization, module declarations).

### CODE REVIEW

**Comment by echeran** [comment]
Also, have people used type aliases with SerDe?  serde::json maps structs to JSON objects, which makes sense, but what if I want to define a locale subtag as `Vec<String>`?

There is also the separate question of whether defining a locale subtag as `Vec<String>` is the right choice -- also open to discussion on that, etc.

**Comment by zbraniecki** [comment]
Would it be useful to add serde serializer/deserializer as a feature for locale like https://github.com/zbraniecki/unic-locale/commit/1450d0ac124bba6ba7e7927f1cd84e10721fa889 ?

File: components/e2etests/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,58 @@
+use serde::{Deserialize, Serialize};
+
+#[derive(Debug, PartialEq, Serialize, Deserialize)]
+enum TestOp {
```
Comment (sffc) [inline components/e2etests/src/lib.rs:4]: Should these go in a common shared crate instead?  #88

Comment (echeran) [reply components/e2etests/src/lib.rs:4]: Ah, I see what you're saying.  Yes, I agree.  I left my comments in the issue.

**Comment by echeran** [comment]
> Would it be useful to add serde serializer/deserializer as a feature for locale like [zbraniecki/unic-locale@1450d0a](https://github.com/zbraniecki/unic-locale/commit/1450d0ac124bba6ba7e7927f1cd84e10721fa889) ?

Hmm, I think I understand, and yes that makes sense.  IIUC, then that would make some of the E2E tests for Locale look like:  JSON test data's input -> SerDe -> structured test data -> runner -> Locale API -> Locale, and then maybe have JSON test data's exp output -> SerDe -> Locale, and then compare.

**Comment by sffc** [comment]
I'm not sure if we want serde to be merely a feature.  It seems like a core piece of functionality when it comes to the data pipeline.

**Comment by echeran** [comment]
I have a commit to make `Locale` serializable just like `LanguageIdentifier` (shameless mimicking for now).  In the process of doing that, I noticed that some of the tests are brittle -- testing error conditions require use the exact error message.  If we have specific error types in the future, then that can be made less brittle.

File: components/e2etests/src/parser/parser_test.rs [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+#[test]
+fn locale_test_output_deser() {
+    let data = r#"
+{"lang": "en",
+ "region": "US",
+ "subtags": [["u", "hc", "buddhist"]]}
+"#;
+
+    let act_test_output: LocaleTestOutput =
+        serde_json::from_str(data).expect("cannot parse sample LocaleTestOutput");
```
Comment (zbraniecki) [inline components/e2etests/src/parser/parser_test.rs:12]: You can use `serde_json::json` macro - https://docs.serde.rs/serde_json/macro.json.html

**Comment by sffc** [comment]
Just revisiting this PR.

Here's one model you could adopt: first, an IterableDataProvider that exports golden test data, and second, a unit test that checks that the current library output is consistent with the golden data.  I expect that you can share most of the code between the golden data generator and the unit test framework.

We should put our heads together and come up with a model that makes sense.  For now, I'm going to switch this PR back to draft.

**Pull request closed without merging**